spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.stats
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import org.apache.spark.sql.delta.actions.{Action, AddFile, FileAction, RemoveFile}
+import org.apache.spark.sql.delta.hooks.AutoCompactPartitionReserve
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+import org.apache.spark.sql.SparkSession
+
+/**
+ * A collector used to aggregate auto-compaction stats for a single commit. The expectation
+ * is to spin this up for a commit and then merging those local stats with the global stats.
+ */
+trait AutoCompactPartitionStatsCollector {
+  def collectPartitionStatsForAdd(file: AddFile): Unit
+  def collectPartitionStatsForRemove(file: RemoveFile): Unit
+  def finalizeStats(tableId: String): Unit
+}
+
+/**
+ * This singleton object collect the table partition statistic for each commit that creates
+ * AddFile or RemoveFile objects.
+ * To control the memory usage, there are `maxNumTablePartitions` per table and 'maxNumPartitions'
+ * partition entries across all tables.
+ * Note:
+ *   1. Since the partition of each table is limited, if this limitation is reached, the least
+ *      recently used table partitions will be evicted.
+ *   2. If all 'maxNumPartitions' are occupied, the partition stats of least recently used tables
+ *      will be evicted until the used partitions fall back below to 'maxNumPartitions'.
+ *   3. The un-partitioned tables are treated as tables with single partition.
+ * @param maxNumTablePartitions The hash space of partition key to reduce memory usage per table.
+ * @param maxNumPartitions The maximum number of partition that can be occupied.
+ */
+class AutoCompactPartitionStats(
+    private var maxNumTablePartitions: Int,
+    private var maxNumPartitions: Int
+) {
```

Comment:
Why on a separate line while the other class `PartitionStat` has them to close the input arguments?! Be consistent 🙏

Reply:
re https://github.com/delta-io/delta/pull/2414#issuecomment-1883922342, perhaps we can address post merge